File: HashmapAndSet/canConstruct.js
const canConstruct = (ransomNote, magazine) => {
  const getCharIndex = (char) => {
    return char.charCodeAt(0) - 97;
  };

  const charCounts = new Array(26).fill(0);
  for (const char of magazine) {
    charCounts[getCharIndex(char)]++;
  }

  for (const char of ransomNote) {
    const i = getCharIndex(char);
    if (--charCounts[i] < 0) return false;
  }

  return true;
};

// Time Complexity: O(n + m), where n is the length of ransomNote 
// and m is the length of magazine.
// Space Complexity: O(1), since the charCounts array 
// has a fixed size of 26.

File: HashmapAndSet/can_construct.py
from collections import Counter


class CanConstruct:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        mag_letters = Counter(magazine)
        for char in ransomNote:
            mag_letters[char] -= 1
            if mag_letters[char] < 0:
                return False
        return True

    # Time complexity: O(n + m) where n is the length of ransomNote and m is the length of magazine.
    # We iterate through the magazine to count the characters, which takes O(m) time.
    # Then we iterate through the ransomNote to check if we have enough characters, which can take O(n) time.
    # Space complexity: O(m) for storing the character counts of the magazine, 
    # though in the worst case it could be O(1) 
    # if we consider the character set to be limited (e.g., only lowercase letters).